source/request-response/subscription_manager.c

```diff
@@ -0,0 +1,627 @@
+
+        struct aws_rr_subscription_listener *listener = iter.element.value;
+        struct aws_rr_subscription_status_event event = {
+            .type = type,
+            .topic_filter = record->topic_filter_cursor,
+            .operation_id = listener->operation_id,
+        };
+
+        (*manager->config.subscription_status_callback)(&event, manager->config.userdata);
+
+        AWS_LOGF_DEBUG(
+            AWS_LS_MQTT_REQUEST_RESPONSE,
+            "request-response subscription manager - subscription event for ('" PRInSTR
+            "'), type: %s, operation: %" PRIu64 "",
+            AWS_BYTE_CURSOR_PRI(record->topic_filter_cursor),
+            s_request_response_subscription_event_type_to_c_str(type),
+            listener->operation_id);
+    }
+}
+
+static int s_rr_activate_idle_subscription(
+    struct aws_rr_subscription_manager *manager,
+    struct aws_rr_subscription_record *record) {
+    int result = AWS_OP_SUCCESS;
+
+    if (manager->is_protocol_client_connected && aws_hash_table_get_entry_count(&record->listeners) > 0) {
+        if (record->status == ARRSST_NOT_SUBSCRIBED && record->pending_action == ARRSPAT_NOTHING) {
+            struct aws_protocol_adapter_subscribe_options subscribe_options = {
+                .topic_filter = record->topic_filter_cursor,
+                .ack_timeout_seconds = manager->config.operation_timeout_seconds,
+            };
+
+            result = aws_mqtt_protocol_adapter_subscribe(manager->protocol_adapter, &subscribe_options);
+            if (result == AWS_OP_SUCCESS) {
+                AWS_LOGF_DEBUG(
+                    AWS_LS_MQTT_REQUEST_RESPONSE,
+                    "request-response subscription manager - initiating subscribe operation for ('" PRInSTR "')",
+                    AWS_BYTE_CURSOR_PRI(record->topic_filter_cursor));
+                record->pending_action = ARRSPAT_SUBSCRIBING;
+            } else {
+                int error_code = aws_last_error();
+                AWS_LOGF_ERROR(
+                    AWS_LS_MQTT_REQUEST_RESPONSE,
+                    "request-response subscription manager - synchronous failure subscribing to ('" PRInSTR
+                    "'), ec %d(%s)",
+                    AWS_BYTE_CURSOR_PRI(record->topic_filter_cursor),
+                    error_code,
+                    aws_error_debug_str(error_code));
+
+                s_emit_subscription_event(manager, record, ARRSET_SUBSCRIPTION_SUBSCRIBE_FAILURE);
+            }
+        }
+    }
+
+    return result;
+}
+
+enum aws_acquire_subscription_result_type aws_rr_subscription_manager_acquire_subscription(
+    struct aws_rr_subscription_manager *manager,
+    const struct aws_rr_acquire_subscription_options *options) {
+    struct aws_rr_subscription_record *existing_record = s_get_subscription_record(manager, options->topic_filter);
+
+    // is no subscription present?
+    if (existing_record == NULL) {
+        s_cull_unused_subscriptions(manager);
```

Comment:
I forgot where we landed on this and can't easily find it in the tech spec but are we going to run the cull every time a new subscription record is requested? Or should we keep unused request-response subs until we need the space?

Reply:
We're doing lazy culling on subscribe regardless of what the tech doc says.

Reply:
As a follow-up, this is not quite accurate anymore.  Subscriptions are purged eagerly and do not linger indefinitely.